### DISCUSSION TITLE
Multithreading parallel executions

### DISCUSSION THREAD

**Discussion by PatrickHuembeli** [jax-ml/jax#11565] [General]
Hi everyone, 

We are running a fairly small neural network (NN) with jax and jit, but we would like to repeat the training many times to evaluate its average performance. Jitting the NN and its gradients takes almost the same amount of time as its training. Therefore, we would like to jit the functions and gradients just once and then run the optimization on several CPUs at the same time. What would be the best way to do this?

I have seen there is the pmap function. So what I was wondering, is it possible to pmap the gradient function `jax.jit(jax.pmap(jax.grad(f(weights))))` and then instead of passing the normal weights we pass a batch of weights? Does this still work normally with the optax optimizers? And is there anything we have to be aware of? Or is there a better solution to this problem?

**Comment by YouJiacheng** [comment]
IIUC, JAX's cpu backend is multi-thread by default.
`jax.pmap` implies `jax.jit`.
Usually you can only see 1 (Abstract) CPU device, so `pmap` doesn't work well with CPU.
"we would like to repeat the training many times to evaluate its average performance. Jitting the NN and its gradients takes almost the same amount of time as its training." - So you can jit once, then run the optimization multiple times, in the same program.
Persistent jit cache is experimental and only available on TPU now.

**Reply by PatrickHuembeli** [reply]
Ok, thanks for the answer. So there is no way of jitting our neural network and its gradients once and then run different optimizations (e.g. with different initializations) on different CPUs without rejitting them? Is this correct?
So basically our issue is, that we want to run the same optimization (with different initialization) let's say a 1000 times. We have 96 CPU cores and want to use each of them for ~10 optimization runs. They way I see it, we have to jit the function on each core separately. Would you agree with that statement?

**Reply by YouJiacheng** [reply]
Ah. You definitely can jit once then run multiple optimizations with different initialization. initialization is just an ordinary operation, there is nothing special.

**Reply by YouJiacheng** [reply]
You can just do:
```
jit
for i in range(1000):
    init
    train_loop
```

**Reply by YouJiacheng** [reply]
This could automatically utilize multiple CPUs, though its efficiency could be lower than.
BTW, if you can abstract the whole process of one training into a jit compatible function map from prngkey to performance. Then you can vmap this function to optimize 1000 NN in parallel on CPUs or GPU.

**Reply by PatrickHuembeli** [reply]
Ah, amazing I didn't know I can jit such a loop. I actually thought this is what one really should avoid. 
Your 2nd answer is even more intriguing and it's what I more or less had in mind in my initial question. So just to be clear, if I have an optimization loop of the form:

```
weights = jax.random.uniform(subkey, shape) 
grad_f = jax.jit(jax.grad(loss, argnums = 0))
for i in range(epochs):
    dw = grad_f(weights)
    weights -= dw*learning_rate
```

I could now just wrap this in a function f(weights, subkeys) and vmap the weights and subkey dimension and feed batches of weights and subkeys? Is this what you are describing above?

**Reply by YouJiacheng** [reply]
I must clarify that my first answer doesn't mean jit the loop. The top jit means jit the train_step.
```
jit the train_step
for i in range(1000):
    init
    train_loop
```

**Reply by PatrickHuembeli** [reply]
Ok, perfect. Thanks a lot for your response.

**Reply by YouJiacheng** [reply]
Second answer mean:
```python
@jax.jit # jit should be the outermost
@jax.vmap
def train_then_eval(key):
    weights = jax.random.uniform(key, shape) 
    grad_f = jax.jit(jax.grad(loss, argnums = 0))
    for i in range(epochs): # could be replaced by lax.scan to accelerate compiling
        dw = grad_f(weights)
        weights -= dw*learning_rate
    return evaluate(weights)

performances = train_then_eval(jax.random.split(jax.random.PRNGKey(0), 1000)) # 1000 performance
# train 1000 NNs in parallel!
```
